flake.nix

```diff
@@ -460,6 +460,9 @@
                 nativeBuildInputs = shellCommonNative.nativeBuildInputs ++ [ pkgs.yarn pkgs.nodejs ];
                 shellHook = ''
                   export FEDIMINT_UI_SHELL=1
+                  cd fedimint-ui
+                  yarn install
+                  yarn build
```

Comment:
This is going to execute every time a dev shell is started, and I bet it takes forever. Dev shell is supposed to init the shell to build in, not start building stuff.

Can we make it another `justfile` command? `just init-yarn` or something?

Reply:
It's really weird that one has to prepare the dev env at all in such a way. A sensible build system should recognize its missing dependencies and just fetch them instead of requiring the user to run a bespoke order of commands to init their environment. 

I also wonder if the dependency directory created by yarn install could instead be a nix package and supplied immutably. We probably need it for packaging anyway.

Reply:
We could wrap `yarn` in a custom script to detect that certain things need to be done one time, leave a marker file etc. then call the actual binary.

A bit (too) magical but possible. Example from some other project where I did that: https://github.com/rustshop/rustshop/blob/cc75f2939a716657e85289e00c2b2de17f718c03/rustshop/flake.nix#L80

Reply:
>  A sensible build system should recognize its missing dependencies and just fetch them instead of requiring the user to run a bespoke order of commands to init their environment.


BTW. I don't know the details, but it looks to me like here it's some kind of self-inflicted problem, due to how the things were glued together.

Reply:
> I don't know the details, but it looks to me like here it's some kind of self-inflicted problem

Maybe someone with more experience in the frontend realm can help here, are we just holding it wrong? @EthnTuttle @wbobeirne

Reply:
> BTW. I don't know the details, but it looks to me like here it's some kind of self-inflicted problem, due to how the things were glued together.

Oh, actually I remember at $previous_dayjob we had similar issues and we were using https://github.com/lerna/lerna for a while. Then switched/combined it with jfrog artifactory.

:shrug: Should have just wrote it all in Rust. ;D

Reply:
Yarn install is quick to call after you've fetched the dependencies once so that should be no problem. But I wouldn't run yarn build, that makes more sense as a just command. You don't even use the built outputs in development, you just run yarn dev.

Reply:
I just noticed that the turborepo config is missing a `dev` command which would likely be the missing piece here. The idea should be that any `packages/ui` project that has a build step also has a `dev` command just like any of the apps, so that if you're editing things in `ui`, they're rebuilt on the fly just as the apps would be. That way you just run `turbo dev` in `fedimint-ui` to work on any of the apps _or_ packages.

Reply:
I've opened #2589 to improve the dev experience so that you don't need to run `yarn build` to do development